### PR TITLE
Job timers 2

### PR DESCRIPTION
**Pull request by metalgearsloth**
Some notes because I picked this PR up and corners were cut
1. Some mind changes might be funky but I got lazy with roles
2. Autosave frequency is 15 minutes in case of server crashes
3. It detects AFK players crappily and tries to not write to DB when people go AFK. If we don't care about server crashes I can just write on round restarts / shutdown and remove autosave entirely.
4. I used semaphoreslims around the role timers dictionary to try and avoid threading issues. It looks like role bans might be susceptible to it if a role ban is added soon enough after someone joins the server but I don't think this has ever come up? Either way timers should okay but babby's first actually threaded code (parallel.for doesn't count).

:cl: Veritius & metalgearsloth
- add: Added role timers. Players will have a 1 week grace period to fufill the requirements for their roles of choice.

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Screenie 
image

File: Resources/Prototypes/Roles/Jobs/Security/warden.yml
Comment (metalgearsloth) [reply Resources/Prototypes/Roles/Jobs/Security/warden.yml:6]: Maybe I should review what veri put in...

**Comment by metalgearsloth** [comment]
image
 I updated the locks to stripeback and it looks a bit better.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (ElectroJr) [inline Content.Server/Roles/RoleTimerSystem.cs:247]: I dunno how best to fix this, because roles are weird & not all roles necessarily have prototypes, but this should really be using prototype IDs. Antags might get renamed, and in the case of Jobs, the job-role name is actually a localized string (its `Role.Name == Loc.GetString(JobPrototype.Name)`).

Maybe explicitly try to cast the `Role` to a `Job` or `TraitorRole` and try to get their relevant prototypes, and then use the prototype IDs? I guess explicitly tracking ObserverRole playtime is also worthwhile.

Or maybe just give the role class an ID field, and have Job & traitor point to the prototype's ID?

**Review by ElectroJr — CHANGES_REQUESTED** [review]
Haven't actually done any testing. Also, I know nothing about DB code.

File: Resources/Locale/en-US/job/role-timers.ftl
Comment (ElectroJr) [inline Resources/Locale/en-US/job/role-timers.ftl:5]: This probably should say more than just "Locked" w/o requiring a tool-tip hover, new players might just think the role is locked on the server.

Also, for the tooltips I think "Require" should be "Requires", but I'm not 100% sure.

File: Content.Server/Administration/Commands/GetRoleTimerCommand.cs
Comment (ElectroJr) [reply Content.Server/Administration/Commands/GetRoleTimerCommand.cs:30]: Actually, having commands for admins to add and remove time is probably quite important?

File: Content.Server/Afk/AFKSystem.cs
Comment (metalgearsloth) [reply Content.Server/Afk/AFKSystem.cs:71]: It's more about having the event issued when they go afk, not up to 10 seconds later. Ideally whenever the event comes in it resets a timer. It's probably fine for now anyway.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (metalgearsloth) [reply Content.Server/Roles/RoleTimerSystem.cs:122]: The AFKSystem will issue the event which should pause their timer now so the comment is less relevant.

File: Content.Server/Station/Systems/StationJobsSystem.cs
Comment (metalgearsloth) [reply Content.Server/Station/Systems/StationJobsSystem.cs:24]: weird, rider wasn't showing it.

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (metalgearsloth) [reply Content.Shared/Roles/SharedRoleTimerSystem.cs:50]: Chucking dependencies onto a prototype always felt dirty to me as you'd need to pass in prototypemanager into it.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (metalgearsloth) [reply Content.Server/Roles/RoleTimerSystem.cs:239]: > Maybe explicitly try to cast the Role to a Job or TraitorRole and try to get their relevant prototypes, and then use the prototype IDs?

Won't that have the same issue of ID overlap? The real issue is roles are bad to work with and I don't have the energy to try and refactor half of mind code trying to deal with it.

Comment (ElectroJr) [reply Content.Server/Roles/RoleTimerSystem.cs:239]: I guess you could somehow extract the prototype name (i.e., `AntagPrototype` -> "antag"), and then use <prototype-name/type>.<prototype-ID> in place of `mindRole.Name`. That's pretty general and should have no overlap?

Or I guess you just need to define a role-prototype that exists just for tracking and give Antag & job roles relevant prototypes? I'm really not sure what the best fix is.

File: Resources/Prototypes/Roles/Jobs/Command/captain.yml
Comment (mirrorcult) [reply Resources/Prototypes/Roles/Jobs/Command/captain.yml:15]: this was discussed by admins and it is intended to be this way

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (ElectroJr) [inline Content.Shared/Roles/SharedRoleTimerSystem.cs:64]: This is still pretty icky. Because there's no yaml linting, there should at least be a test that ensures that every job prototype has a corresponding "Job<ID>" role prototype.

File: Content.Server/Administration/Commands/RoleTimers/AddOverallTimeCommand.cs
Comment (ElectroJr) [inline Content.Server/Administration/Commands/RoleTimers/AddOverallTimeCommand.cs:36]: Either this should be `FromMinutes`, or the variable, error & help/usage need to be updated.

File: Content.Shared/Audio/ContentAudioSystem.cs
Comment (metalgearsloth) [inline Content.Shared/Audio/ContentAudioSystem.cs:14]: A my branch got cooked, will revert before merge.

File: Content.Server/Database/ServerDbBase.cs
Comment (PJB3005) [inline Content.Server/Database/ServerDbBase.cs:401]: Why does this return a nullable?

File: Content.Server/Database/ServerDbBase.cs
Comment (PJB3005) [inline Content.Server/Database/ServerDbBase.cs:413]: Ditto nullable

File: Content.Server/GameTicking/GameTicker.RoundFlow.cs
Comment (PJB3005) [inline Content.Server/GameTicking/GameTicker.RoundFlow.cs:196]: I don't like the way this is handled, but at the same time I see some other code already started doing this so...

This should be handled the same way as player preferences: don't allow people to even ready up before database info is loaded.

File: Content.Server/GameTicking/GameTicker.Spawning.cs
Comment (PJB3005) [inline Content.Server/GameTicking/GameTicker.Spawning.cs:90]: Bad style change.

File: Content.Server/Roles/RoleTimerManager.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerManager.cs:129]: So two things here:
1. All the actual logic in the critical sections is so small you're wasting more CPU cycles using asynchronous semaphore awaits than if you'd just use `lock()`. Something which would also avoid all the asynchronous leakage into the consuming code.
2. This synchronization logic isn't even necessary because the sync context ensures awaits resume on the main thread.
3. For future advice: if you really really need to lock something, you should still have put the synchronization externally from these internal methods. When loading role timers you are constantly grabbing and releasing the lock. Not good.

File: Content.Client/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Client/Roles/RoleTimerSystem.cs:21]: This is out-of-simulation info, it should be managed on a regular manager via `NetMessage`s, not entity messages.

File: Content.Server/Roles/RoleTimerManager.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerManager.cs:45]: As stated before, this should be sent with net messages and out-of-simulation stuff directly.

File: Content.Server/Roles/RoleTimerManager.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerManager.cs:48]: Returning false if the player doesn't exist feels error prone. That should maybe throw an error instead.

File: Content.Server/Roles/RoleTimerManager.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerManager.cs:96]: Please I beg of you remove all these same-line if statements.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:34]: Do not set default values that will be replaced by config read.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:107]: Entity systems should not be in charge of making sure that things are correctly saved on shutdown. Such data management operations should be relegated to managers.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:159]: There's nothing here to enforce the database writes are properly committed before the server shuts down.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:74]: This can just be `return new List<string>(GetTimedRoles(contentData.Mind));`. Or return an enumerable directly.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:85]: Full database commit every time somebody changes their roles slightly seems excessive. This is going to cause a massive spike of DB commits on something like round restart.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:97]: Ditto about DB spam.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:115]: This should be handled by the role timer manager itself.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:121]: Ditto, role timer manager ought to do this.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:216]: Look it's 1:30 AM here but these functions are *way* too similar so something is wrong here.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [reply Content.Server/Roles/RoleTimerSystem.cs:85]: The roles are already removed from the mind's role list by the time this event is raised. This is partially the fault of the event not being named in past tense.

Either way, this code is incorrect. `Save()` won't find the role because it's already not in the list anymore, so some roles will fail to be accounted for.

Comment (PJB3005) [reply Content.Server/Roles/RoleTimerSystem.cs:85]: I would argue that role timer info should use **real time**, not simulation timing. This is mostly a player facing matter, so punishing players for any server timer disparities seems wrong.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:145]: This should not use simulation time or `Update()`. If the game server pauses, necessary autosaves may not run.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:214]: ALL of these operations generate **individual** DB transactions with no batching of any kind. Per player. This is **not** acceptable.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:267]: `jobs` does not need to be passed by `ref`

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:269]: I'm pointing it out again because it pissed me off: no one-line ifs.

File: Content.Server/Roles/RoleTimerSystem.cs
Comment (PJB3005) [inline Content.Server/Roles/RoleTimerSystem.cs:305]: No `.Result`

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (PJB3005) [inline Content.Shared/Roles/SharedRoleTimerSystem.cs:44]: This lazy initialization logic is still insane because it's all just cached data you're pulling in here anyways.

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (PJB3005) [inline Content.Shared/Roles/SharedRoleTimerSystem.cs:76]: Formatting in Fluent

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (PJB3005) [inline Content.Shared/Roles/SharedRoleTimerSystem.cs:86]: Formatting in Fluent

File: Content.Shared/Roles/SharedRoleTimerSystem.cs
Comment (PJB3005) [inline Content.Shared/Roles/SharedRoleTimerSystem.cs:101]: Formatting in Fluent

File: Resources/Locale/en-US/job/department.ftl
Comment (PJB3005) [inline Resources/Locale/en-US/job/department.ftl:7]: These should probably be capitalized.

File: Resources/Locale/en-US/job/job-names.ftl
Comment (PJB3005) [inline Resources/Locale/en-US/job/job-names.ftl:45]: These do not follow normal fluent naming at all

**Review by PJB3005 — CHANGES_REQUESTED** [review]
I do not like how timing information is tracked via polling. I have explained before that it is quite easy to get exact timing information by treating all playtime-and-AFK related actions as a sequence of timed state changes, but I don't have the energy to block the PR over it.

The database and caching logic however is a deal breaker: this needs a rewrite.

**Comment by metalgearsloth** [comment]
Ike > Veritius > Sloth > ???

**Comment by Emisse** [comment]
pleae..,,. i am goingto cry now

**Comment by PJB3005** [comment]
#9978

**Pull request closed without merging**
